### PR TITLE
SugaredLogger: Turn error into zap.Error 

### LINKED ISSUE #1184: SugaredLogger: Turn error into zap.Error
**Is your feature request related to a problem? Please describe.**

With Zap's sugared logger, if you provide a plain error object to Infow/Errorw/etc.,
it does not desugar the error into a zap.Field.

```go
log.Warnw("something happened", err) // bad

log.Warnw("something happened", "err", err) // good
// desugars to
log.Desugar().Warn("something happened", zap.NamedError("err", err))
```

This is inconvenient for the majority case where there will be only one error.

**Describe the solution you'd like**

We should turn the first case into,

```go
log.Desugar().Warn("something happened", zap.Error(err))
```

We could even be smart about it in desugarFields
and make sure there's only one instance of "err".
Others have to have a field name.

**Describe alternatives you've considered**

Add "err" next to every err argument.

**Is this a breaking change?**
No

### PR DESCRIPTION
**Pull request by craigpastro**
Resolves #1184.

Please let me know what you think! Thanks.

### CODE REVIEW

**Review by abhinav — APPROVED** [review]
This looks great! Thanks!

@mway @sywhang what do you think?

**Review by sywhang — APPROVED** [review]
This looks good to me. Thanks for doing this work!

File: sugar.go [hunk 1]
```diff
@@ -340,6 +341,7 @@ func (s *SugaredLogger) sweetenFields(args []interface{}) []Field {
 	// fields, we shouldn't penalize them with extra allocations.
 	fields := make([]Field, 0, len(args))
 	var invalid invalidPairs
+	var seenError bool
```
Comment (mway) [inline sugar.go:344]: nit: let's combine these into a `var` block

File: sugar_test.go [hunk 2]
```diff
@@ -198,9 +214,10 @@ func TestSugarStructuredLogging(t *testing.T) {
 	}
 
 	// Common to all test cases.
+	err := errors.New("qux")
 	context := []interface{}{"foo", "bar"}
-	extra := []interface{}{"baz", false}
-	expectedFields := []Field{String("foo", "bar"), Bool("baz", false)}
+	extra := []interface{}{err, "baz", false}
+	expectedFields := []Field{String("foo", "bar"), Error(err), Bool("baz", false)}
```
Comment (mway) [inline sugar_test.go:220]: nit: let's combine these into a `var` block here as well

**Review by mway — APPROVED** [review]
This is great, thanks! Couple of style nits, but otherwise lgtm.

File: sugar.go (hunk 1 above)
Comment (craigpastro) [reply sugar.go:344]: Sure! Thank you. Done.

File: sugar_test.go (hunk 2 above)
Comment (craigpastro) [reply sugar_test.go:220]: Combined into a var block here too. Thank you!

**Pull request merged**